src/librustc_typeck/outlives/implicit_infer.rs

```diff
@@ -0,0 +1,244 @@
+    predicates_added: bool,
+}
+
+impl<'cx, 'tcx> ItemLikeVisitor<'tcx> for InferVisitor<'cx, 'tcx> {
+    fn visit_item(&mut self, item: &hir::Item) {
+        let item_did = self.tcx.hir.local_def_id(item.id);
+
+        let node_id = self.tcx
+            .hir
+            .as_local_node_id(item_did)
+            .expect("expected local def-id");
+        let item = match self.tcx.hir.get(node_id) {
+            hir::map::NodeItem(item) => item,
+            _ => bug!(),
+        };
+
+        let mut local_predicate_map = FxHashMap();
+        match item.node {
+            hir::ItemUnion(..) |
+            hir::ItemEnum(..) |
+            hir::ItemStruct(..) => {
+
+                let adt_def = self.tcx.adt_def(item_did);
+
+                // Iterate over all fields in item_did
+                for field_def in adt_def.all_fields() {
+
+                    // Calculating the predicate requirements necessary
+                    // for item_did.
+                    //
+                    // For field of type &'a T (reference) or TyAdt
+                    // (struct/enum/union) there will be outlive
+                    // requirements for adt_def.
+                    let field_ty = self.tcx.type_of(field_def.did);
+                    let requirements_for_item_did = required_predicates_to_be_wf(
+                        self.tcx,
+                        field_ty,
+                        self.global_inferred_outlives,
+                     ).into_iter().collect();
+
+                    // Add the requirements_for_item_did to the local map
+                    local_predicate_map.insert(
+                        item_did,
+                        Rc::new(requirements_for_item_did),
+                    );
+                }
+            },
+
+            _ => {},
+        };
+
+        if local_predicate_map.len() > 0 {
+            self.predicates_added = true;
+            self.global_inferred_outlives
+                .extend(local_predicate_map);
+        }
+    }
+
+    fn visit_trait_item(&mut self, trait_item: &'tcx hir::TraitItem) {}
+
+    fn visit_impl_item(&mut self, impl_item: &'tcx hir::ImplItem) {}
+}
+
+fn required_predicates_to_be_wf<'tcx>(
+    tcx: TyCtxt<'_, 'tcx, 'tcx>,
+    field_ty: Ty<'tcx>,
+    global_inferred_outlives: &FxHashMap<DefId, Rc<Vec<ty::Predicate<'tcx>>>>,
+) -> HashSet<ty::Predicate<'tcx>> {
+
+    let mut required_predicates = HashSet::new();
+    match field_ty.sty {
+
+        // The field is of type &'a T which means that we will have
+        // a prediate requirement of T: 'a (T outlives 'a).
+        //
+        // We also want to calculate potential predicates for the T
+        ty::TyRef(reg, mt) => {
+            let outlives_pred = ty::Binder(ty::OutlivesPredicate(mt.ty, reg))
+                .to_predicate();
+            required_predicates.insert(outlives_pred);
+
+            let predcates_for_mt: HashSet<ty::Predicate<'tcx>> =
```

Comment:
@nikomatsakis According to your comment it was necessary to recurse on the type `mt`. However wouldnt the `ItemLikeVisitor` do this already?